components/script/dom/htmlinputelement.rs

```diff
@@ -584,8 +627,30 @@ impl VirtualMethods for HTMLInputElement {
                             &atom!("password") => InputType::InputPassword,
                             _ => InputType::InputText,
                         };
-                        self.input_type.set(value);
-                        if value == InputType::InputRadio {
+
+                        // https://html.spec.whatwg.org/multipage/#input-type-change
+                        let old_idl_value = self.Value();
+                        self.input_type.set(new_type);
+
+                        let old_value_mode = HTMLInputElement::get_value_mode_for_input_type(old_type);
+                        let new_value_mode = HTMLInputElement::get_value_mode_for_input_type(new_type);
+
+                        if old_value_mode == ValueMode::Value && old_idl_value != DOMString::from("")
```

Comment:
This could really use a match-expression against `(old_value_mode, old_idl_value, new_value_mode)`.